synapse/lib/stormlib/vault.py

```diff
@@ -0,0 +1,648 @@
+
+        if not check and not self.runt.asroot:
+            iden = vault.get('iden')
+            mesg = f'Insufficient permissions for user {self.runt.user.name} to vault {iden}.'
+            raise s_exc.AuthDeny(mesg=mesg, user=self.runt.user)
+
+    async def _addVault(self, name, vtype, scope, iden, data):
+        name = await s_stormtypes.tostr(name)
+        vtype = await s_stormtypes.tostr(vtype)
+        scope = await s_stormtypes.tostr(scope, noneok=True)
+        iden = await s_stormtypes.tostr(iden, noneok=True)
+        data = await s_stormtypes.toprim(data)
+
+        user = None
+        if not self.runt.asroot:
+            user = self.runt.user
+
+        return await self.runt.snap.core.addVault(name, vtype, scope, iden, data, user=user)
+
+    async def _getVault(self, name):
+        vault = self.runt.snap.core.getVaultByIden(name)
+        if not vault:
+            vault = self.runt.snap.core.reqVaultByName(name)
+
+        self._reqEasyPerm(vault, s_cell.PERM_READ)
+
+        iden = vault.get('iden')
+        return Vault(self.runt, iden)
+
+    async def _getByName(self, name):
+        name = await s_stormtypes.tostr(name)
+
+        vault = self.runt.snap.core.reqVaultByName(name)
+        self._reqEasyPerm(vault, s_cell.PERM_READ)
+
+        iden = vault.get('iden')
+        return Vault(self.runt, iden)
+
+    async def _getByIden(self, viden):
+        viden = await s_stormtypes.tostr(viden)
+
+        vault = self.runt.snap.core.reqVaultByIden(viden)
+        self._reqEasyPerm(vault, s_cell.PERM_READ)
+
+        return Vault(self.runt, viden)
+
+    async def _getByType(self, vtype, scope=None):
+        vtype = await s_stormtypes.tostr(vtype)
+        scope = await s_stormtypes.tostr(scope, noneok=True)
+
+        vault = self.runt.snap.core.getVaultByType(vtype, self.runt.user.iden, scope)
+        self._reqEasyPerm(vault, s_cell.PERM_READ)
+
+        iden = vault.get('iden')
+        return Vault(self.runt, iden)
+
+    async def _listVaults(self):
+        for vault in self.runt.snap.core.listVaults(user=self.runt.user):
+            yield vault
+
+    async def _setDefault(self, vtype, scope):
+        vtype = await s_stormtypes.tostr(vtype)
+        scope = await s_stormtypes.tostr(scope, noneok=True)
+
+        user = None
+        if not self.runt.asroot:
+            user = self.runt.user
+
+        return await self.runt.snap.core.setVaultDefault(vtype, scope, user=user)
+
+@s_stormtypes.registry.registerType
+class VaultData(s_stormtypes.Prim):
+    '''
+    Implements the Storm API for Vault data.
+
+    Callers (instantiation) of this class must have already checked that the user has at least
+    PERM_EDIT to the vault.
+    '''
+    _storm_typename = 'vault:data'
+    _ismutable = False
+
+    def __init__(self, runt, valu, path=None):
+        s_stormtypes.Prim.__init__(self, valu, path=path)
+        self.runt = runt
+
+    async def setitem(self, name, valu):
```

Comment:
This would be the third setitem() implementation which needs a readonly runtime check.

We should probably make a decorator for the `setitem()` checks we can easily apply.

Reply:
This should get resolved with #3364